### PR TITLE
Include the effect of the sources in the reference state

### PR DESCRIPTION
**Pull request by maxpkatz**
## PR summary

As discussed in #1645 it appears to be incorrect that we include the source term in the characteristic tracing but do not include the source term in the reference state. (This is done for PPM only.)

## PR checklist

- [x] test suite needs to be run on this PR
- [x] this PR will change answers in the test suite to more than roundoff level
- [ ] all newly-added functions have docstrings as per the coding conventions
- [ ] the `CHANGES` file has been updated, if appropriate
- [ ] if appropriate, this change is described in the docs

### CODE REVIEW

File: Source/hydro/trace_ppm.cpp [hunk 1]
```diff
@@ -468,11 +490,11 @@ Castro::trace_ppm(const Box& bx,
 
       // Set the reference state
       // This will be the fastest moving state to the right
-      Real rho_ref = Ip_rho[2];
-      Real un_ref = Ip_un_2;
+      Real rho_ref = Ip_rho[2] + hdt * Ip_src_un_2;
```
Comment (zingale) [inline Source/hydro/trace_ppm.cpp:493]: this should be Ip_src_rho_2

**Comment by zingale** [comment]
I like this approach.  

I will want to test a longer run of `flame_wave` just to see if there are any issues

**Comment by zingale** [comment]
tests: http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2021-05-19-002/index.html

**Comment by zingale** [comment]
I ran flame_wave for 47000 steps, comparing the current development to this PR

Here's development:
flame_wave_1000Hz_development_smallplt47484_slice

here's this PR:
flame_wave_1000Hz_source_reference_smallplt47140_slice

they look largely the same.  Maybe a little more stuff at the symmetry axis with this PR.  But the flame structure looks the same and they are taking pretty much the same timestep as before (and as we did in paper II).

**Comment by zingale** [comment]
Note: this change eliminates the warnings like:
```
WARNING: (rho e)_r < 0 or pr < small_pres in Riemann: -2.580387688e+17 5.195193164e+10 1e-200
WARNING: (rho e)_r < 0 or pr < small_pres in Riemann: -2.581237387e+17 5.594825754e+10 1e-200
```
in the `DustCollapse` problem during the first timestep when using ``inputs_3d_poisson_regtest``

**Comment by maxpkatz** [comment]
I have also observed this in some cases in the giant impact problem (#1650) and this change resolves the issue there too.

**Comment by zingale** [comment]
can we do the ppm_integrate_profile instead of ppm_integrate_profile_single?  then I think it is ready to go.

**Comment by maxpkatz** [comment]
I think I am doing more work than I need to here -- I added Im_src_rho_2 and Ip_src_rho_0, but these represent the waves moving in the other direction of the interface so they don't participate in the reference state. So maybe I just delete those extraneous ones and use ppm_int_profile_single since there's only 2 out of 3 needed?

**Comment by zingale** [comment]
that's fine

**Comment by zingale** [comment]
oh, but I think ppm_int_profile_single is per wave.  So you can't only get Ip for the 3 wave you also get Im for the 3 wave

**Comment by maxpkatz** [comment]
Yes, that must be why I did that. OK, updated to use ppm_int_profile for rho and rhoe.

**Review by zingale — APPROVED** [review]

**Comment by zingale** [comment]
tests: http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2021-06-03-004/index.html

the rotating star w/ NSE gets an FPE now :/

**Pull request merged**